README
Lyrics plugins for ncmpc
============================



CHANGES
08.01.2016:
- removed 01-lyricwiki.rm script cause it stopped working
- instead replaced it by 01-lyricwiki.py


01-lyricwiki.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

#  Load lyrics from lyrics.wikia.com
#  Author: Eric Scheibler   (email[at]eric-scheibler[dot]de)
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

import requests, argparse, sys
from xml.dom import minidom
from subprocess import Popen, PIPE, STDOUT

def download_page(url):
    r = requests.get(url)
    r.encoding = "utf-8"
    return r.text.encode("utf-8")


def extract_url_from_xml_data(xml_contents):
    xml = minidom.parseString(xml_contents)
    for url_node in xml.getElementsByTagName('url'):
        for url in url_node.childNodes:
            return url.nodeValue

def html_to_text(html_content):
    p = Popen(["pandoc", "-f", "html", "-t", "plain"], stdout=PIPE, stdin=PIPE, stderr=STDOUT)
    return p.communicate(input=html_content)[0]


# create parser object
parser = argparse.ArgumentParser(description="Lyrics parser for http://lyrics.wikia.com")
parser.add_argument("artist", nargs="?", default="", help="Song artist")
parser.add_argument("title", nargs="?", default="", help="Song title")
args = parser.parse_args()
if args.artist == "":
    print "Song artist missing"
    sys.exit(1)
if args.title == "":
    print "Song title missing"
    sys.exit(1)

# query url of actual lyrics page
url = "http://lyrics.wikia.com/api.php?action=lyrics&fmt=xml&func=getSong&artist=%s&song=%s" \
        % (requests.utils.quote(args.artist), requests.utils.quote(args.title))
url = extract_url_from_xml_data(download_page(url))
if not bool(url):
    print("Found no url")
    sys.exit(1)
elif "action=edit" in url:
    print("Found no result")
    sys.exit(69)

# get lyrics
lyric_lines = []
song_part_found = False
for line in html_to_text(download_page(url)).split("\n"):
    if song_part_found:
        if "External links" in line:
            break
        else:
            lyric_lines.append(line)
    else:
        if "music Gracenote" in line:
            song_part_found = True
print("%s - %s\n%s\n%s" \
        % (args.artist, args.title, '-' * (len(args.artist) + 3 + len(args.title)), '\n'.join(lyric_lines)))


01-lyricwiki.rb
#!/usr/bin/env ruby
# encoding: utf-8

#  (c) 2004-2008 The Music Player Daemon Project
#  http://www.musicpd.org/
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

#
# Load lyrics from lyrics.wikia.com, formerly lyricwiki.org
#

require 'uri'
require 'net/http'
require 'cgi'
require 'iconv'

url = "http://lyrics.wikia.com/api.php?action=lyrics&fmt=xml&func=getSong" + \
    "&artist=#{URI.escape(ARGV[0])}&song=#{URI.escape(ARGV[1])}"
response = Net::HTTP.get(URI.parse(url))

if not response =~ /<url>\s*(.*?)\s*<\/url>/im
	$stderr.puts "No URL in response!"
	exit(1)
end

url = $1
exit(69) if url =~ /action=edit$/

response = Net::HTTP.get(URI.parse(url))
if not response =~ /<div class='lyricbox'>\s*(.*?)\s*<!--/im
	$stderr.puts "No <div class='lyricbox'> in lyrics page!\n"
	exit(1)
end

if not $1 =~ /^.*<\/div>(.*?)$/im
	$stderr.puts "Couldn't remove leading XML tags in lyricbox!\n"
	exit(1)
end

# lyrics come in Latin1, but we need UTF-8
lyrics_latin1 = CGI::unescapeHTML($1.gsub(/<br \/>/, "\n"))
lyrics_utf8 = Iconv.conv('UTF-8//TRANSLIT//IGNORE', 'Latin1', lyrics_latin1)

# remove java script code from lyrics
lyrics_utf8 = lyrics_utf8.sub(/.*<\/script>/, '')

# replace unicode chars
for regexp in lyrics_utf8.to_enum(:scan, /&#([0-9]+);/).map { Regexp.last_match } do
    lyrics_utf8 = lyrics_utf8.gsub(regexp[0], regexp[1].to_i.chr('UTF-8'))
end

puts lyrics_utf8


01-dark-lyrics-com.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import requests, argparse, sys
from subprocess import Popen, PIPE, STDOUT

base_url = "http://www.darklyrics.com"

def download_page(url):
    r = requests.get(url)
    r.encoding = "utf-8"
    p = Popen(["pandoc", "-f", "html", "-t", "markdown"], stdout=PIPE, stdin=PIPE, stderr=STDOUT)
    return p.communicate(input=r.text.encode("utf-8"))[0]

parser = argparse.ArgumentParser(description="Lyrics parser for www.darklyrics.com")
parser.add_argument("artist", nargs="?", default="", help="Song artist")
parser.add_argument("title", nargs="?", default="", help="Song title")
args = parser.parse_args()
if args.artist == "":
    print "Song artist missing"
    sys.exit(1)
if args.title == "":
    print "Song title missing"
    sys.exit(1)

file = open("/tmp/lyrics_out", "a")
file.write("%s -- %s\n" % (args.artist, args.title))
file.close()

# load results page
result_page = download_page("%s/search?q=%s+%s" % (base_url, \
        args.artist.replace(" ","+"), args.title.replace(" ","+")))
song_url = ""
song_id_on_page = ""
song_part_found = False
for line in result_page.split("\n"):
    if line.startswith("### Songs:"):
        song_part_found = True
        continue
    if song_part_found and "](" in line:
        song_artist = line.strip()[1:].split("](")[0].strip().split(" - ")[0]
        if song_artist.lower() == args.artist.lower():
            song_url = line[:line.__len__()-1].split("](")[1]
            song_id_on_page = song_url.split("#")[-1]
            break
if song_url == "":
    sys.exit(2)

# parse songtext website
result_page = download_page("%s/%s" % (base_url, song_url))
song_name = ""
lyrics = ""
lyrics_start = False
empty_line_detected = False
for line in result_page.split("\n"):
    if song_name == "" and \
            "(#%s)" % song_id_on_page in line:
        song_name = line.split("](")[0].replace("[","").strip()
        lyrics = "%s\n" % song_name
        lyrics += "-" * song_name.__len__() + "\n"
        continue
    if "# %s" % song_name in line:
        lyrics_start = True
        continue
    if lyrics_start and ( \
            "### " in line or \
            "Thanks to " in line or \
            "Submits, comments" in line):
        break
    if lyrics_start:
        line = line.replace("\\", "").strip()
        if line == "":
            if empty_line_detected == False:
                lyrics += "\n"
                empty_line_detected = True
        else:
            lyrics += "%s\n" % line
            empty_line_detected = False

print lyrics[:lyrics.__len__()-1]
sys.exit(0)


03-lyricsmania-com.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

#  Load German lyrics from lyricsmania.com
#  Author: Eric Scheibler   (email[at]eric-scheibler[dot]de)
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#


import requests, argparse, sys, re
from subprocess import Popen, PIPE, STDOUT

base_url = "http://www.lyricsmania.com"

def download_page(url):
    r = requests.get(url)
    r.encoding = "utf-8"
    p = Popen(["pandoc", "-f", "html", "-t", "markdown"], stdout=PIPE, stdin=PIPE, stderr=STDOUT)
    return p.communicate(input=r.text.encode("utf-8"))[0]

parser = argparse.ArgumentParser(description="Lyrics parser for www.darklyrics.com")
parser.add_argument("artist", nargs="?", default="", help="Song artist")
parser.add_argument("title", nargs="?", default="", help="Song title")
args = parser.parse_args()
if args.artist == "":
    print "Song artist missing"
    sys.exit(1)
if args.title == "":
    print "Song title missing"
    sys.exit(1)

# load results page
result_page = download_page("%s/searchnew.php?k=%s+%s" % (base_url, \
        args.artist.replace(" ","+"), args.title.replace(" ","+")))
song_url = ""
song_id_on_page = ""
song_part_found = False
for line in result_page.splitlines():
    if "Z.html" in line:
        song_part_found = True
        continue
    if song_part_found and "](" in line:
        song_url = line.split("](")[1].split(" ")[0]
        break
if song_url == "":
    sys.exit(2)

# parse songtext website
result_page = download_page("%s/%s" % (base_url, song_url))
lyrics = ""
lyrics_start = False
empty_line_detected = False
for line in result_page.splitlines():
    if line.startswith("### "):
        if args.artist.lower() != line[4:].lower():
            sys.exit(3)
    if "Lyrics to" in line:
        lyrics_start = True
        lyrics = "%s\n" % re.sub(r'\*\*.*\*\*(.*)', r'\1', line).strip()
        continue
    if lyrics_start and ( \
            "#####" in line or \
            line.startswith("[")):
        break
    if lyrics_start:
        line = line.replace("\\", "").strip()
        if line == "":
            if empty_line_detected == False:
                lyrics += "\n"
                empty_line_detected = True
        else:
            lyrics += "%s\n" % line
            empty_line_detected = False

print lyrics[:lyrics.__len__()-1]
sys.exit(0)


02-dark-lyrics-com.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

#  Load metal lyrics from darklyrics.com
#  Author: Eric Scheibler   (email[at]eric-scheibler[dot]de)
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#

import requests, argparse, sys
from subprocess import Popen, PIPE, STDOUT

base_url = "http://www.darklyrics.com"

def download_page(url):
    r = requests.get(url)
    r.encoding = "utf-8"
    p = Popen(["pandoc", "-f", "html", "-t", "markdown"], stdout=PIPE, stdin=PIPE, stderr=STDOUT)
    return p.communicate(input=r.text.encode("utf-8"))[0]

parser = argparse.ArgumentParser(description="Lyrics parser for www.darklyrics.com")
parser.add_argument("artist", nargs="?", default="", help="Song artist")
parser.add_argument("title", nargs="?", default="", help="Song title")
args = parser.parse_args()
if args.artist == "":
    print "Song artist missing"
    sys.exit(1)
if args.title == "":
    print "Song title missing"
    sys.exit(1)

# load results page
result_page = download_page("%s/search?q=%s+%s" % (base_url, \
        args.artist.replace(" ","+"), args.title.replace(" ","+")))
song_url = ""
song_id_on_page = ""
song_part_found = False
for line in result_page.split("\n"):
    if line.startswith("### Songs:"):
        song_part_found = True
        continue
    if song_part_found and "](" in line:
        song_artist = line.strip()[1:].split("](")[0].strip().split(" - ")[0]
        if song_artist.lower() == args.artist.lower():
            song_url = line[:line.__len__()-1].split("](")[1]
            song_id_on_page = song_url.split("#")[-1]
            break
if song_url == "":
    sys.exit(2)

# parse songtext website
result_page = download_page("%s/%s" % (base_url, song_url))
song_name = ""
lyrics = ""
lyrics_start = False
empty_line_detected = False
for line in result_page.split("\n"):
    if song_name == "" and \
            "(#%s)" % song_id_on_page in line:
        song_name = line.split("](")[0].replace("[","").strip()
        lyrics = "%s\n" % song_name
        lyrics += "-" * song_name.__len__() + "\n"
        continue
    if "# %s" % song_name in line:
        lyrics_start = True
        continue
    if lyrics_start and ( \
            "### " in line or \
            "Thanks to " in line or \
            "Submits, comments" in line):
        break
    if lyrics_start:
        line = line.replace("\\", "").strip()
        if line == "":
            if empty_line_detected == False:
                lyrics += "\n"
                empty_line_detected = True
        else:
            lyrics += "%s\n" % line
            empty_line_detected = False

print lyrics[:lyrics.__len__()-1]
sys.exit(0)
